demos/object_detection_demo_ssd_async/main.cpp

```diff
@@ -354,58 +403,194 @@ int main(int argc, char *argv[]) {
-            if (isAsyncMode) {
-                async_infer_request_curr.swap(async_infer_request_next);
-            }
+                    request = std::move(emptyRequests.front());
+                    emptyRequests.pop_front();
+                }
+                frameToBlob(frame, request, imageInputName);
+                
+                bool frameMode = isUserSpecifiedMode;
+                request->SetCompletionCallback([&mutex,
+                                                &completedRequestResults,
+                                                nextFrameId,
+                                                frame,
+                                                request,
+                                                &outputName,
+                                                startTime,
+                                                frameMode,
+                                                &isUserSpecifiedMode,
+                                                &emptyRequests,
+                                                &callbackException,
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::unique_ptr<LockedMemory<const void>>(new
+                                        LockedMemory<const void>(as<MemoryBlob>(request->GetBlob(outputName))->rmap())),
+                                    startTime,
+                                    frameMode == isUserSpecifiedMode
+                                }));
+                            
+                            if (isUserSpecifiedMode == frameMode) {
+                                emptyRequests.push_back(std::move(request));
+                            }
+                        }
+                        catch (...) {
+                            if (!callbackException) {
+                                callbackException = std::current_exception();
+                            }
+                        }
+                    }
+                    condVar.notify_one();
+                });
 
-            const int key = cv::waitKey(1);
-            if (27 == key)  // Esc
-                break;
-            if (9 == key) {  // Tab
-                isAsyncMode ^= true;
-                isModeChanged = true;
+                request->StartAsync();
+                nextFrameId++;
             } else {
-                presenter.handleKey(key);
+                std::unique_lock<std::mutex> lock(mutex);
+
+                while (callbackException == nullptr && emptyRequests.empty() && completedRequestResults.empty()) {
+                    condVar.wait(lock);
+                }
```

Comment:
This seems odd. You wait for one of these objects to be updated from a background thread, but then you immediately release the lock and access those objects while the lock is _not_ held - isn't that thread-unsafe?

Reply:
I've replaced `while` with `if` because `while` is not actually useful here, as experiments have shown.

security_barrier_camera_demo and classification_demo might need to be fixed too, because it looks like they have similar problem:
```
std::unique_lock<std::mutex> lk(tasksMutex);
while (running && tasks.empty()) {
        tasksCondVar.wait(lk);
}
```

```
{
        std::unique_lock<std::mutex> lock(mutex);

        while (!irCallbackException
                  && emptyInferRequests.empty() && completedInferRequests.empty()) {
                condVar.wait(lock);
        }
}
```

Reply:
> I've replaced while with if because while is not actually useful here, as experiments have shown.

No, `while` is necessary because of spurious wakeups.

Reply:
You haven't really fixed it yet. I can see that both `callbackException` and `emptyRequests` are accessed in the main loop while the mutex is not held.